### ISSUE TITLE
skillopt's edit budget L is advisory only: nothing compares applied edits against it (observed applied_changes:6 under L:4)

### ISSUE THREAD

**Issue by OsherElhadad** [skillberry-ai/cap-evolve#382]
## Summary

SkillOpt's "textual learning rate" — the per-step edit budget `L` that decays on a `constant|linear|cosine` schedule — is **advisory only**. It is injected into the optimizer prompt as a sentence and recorded in telemetry, but nothing ever compares the number of edits actually applied against it, and nothing rejects or penalizes a step that exceeds it.

Observed live: a step logged `applied_changes: 6` under `edit_budget: 4`, with no consequence.

So the algorithm's headline mechanism is a request, not a constraint. That matters because the whole framing — annealing an edit budget the way a learning rate is annealed — implies the budget binds. If the model ignores it, the schedule is decorative and the `linear`/`cosine` options are choosing between two numbers that never take effect.

## Evidence

`L` reaches the model as prose (`core/cap_evolve/skillopt.py:115`):

```python
f"You may make **at most L = {edit_budget}** bounded edits this step "
```

It is then recorded as telemetry (`skillopt.py:337-338`, `:344-345`):

```python
global_step=global_step, edit_budget=L,
requested_edits=L, applied_changes=applied,
...
step["edit_budget"] = L
step["applied_changes"] = applied
```

`grep -n "edit_budget\|applied_changes\|requested_edits" core/cap_evolve/skillopt.py` returns only the
schedule construction, the prompt string, and these telemetry writes. There is no comparison of
`applied` to `L`, no clamp, no rejection, and no gate input derived from the difference. The schedule
itself is real and correctly clamped to integers (`lr_schedule.py:42-55`) — the bug is that its output
does not constrain anything.

## Why it was hard to notice

Two reasons, both instances of patterns already tracked in #365:

1. **The telemetry records the violation without acting on it.** `applied_changes: 6` next to
   `edit_budget: 4` is written to the event stream every step, so the data proving the budget does not
   bind has been there all along — nothing reads it.
2. **The `mock` optimizer never reads the prompt.** All cheap testing uses it, so no deterministic test
   could ever observe the model exceeding a prompt-stated budget. This is the same blind spot that hid
   #371, #109 and #111.

## What to decide, before writing code

This needs a decision about intent, not just a patch:

- **If `L` is meant to bind**, then a step whose applied-edit count exceeds it should be refused, or the
  excess edits dropped, or the candidate marked indecisive — and there should be a test asserting a
  6-edit candidate under `L=4` does not silently proceed. Note the honest counter-argument: hard-refusing
  a good candidate for exceeding an arbitrary edit count could reject real improvements, and the val gate
  already protects against overshoot. So "bind it" is not obviously right.
- **If `L` is meant to be advisory**, then say so plainly in the skill and stop calling it a learning
  rate, because the analogy is what creates the expectation that it binds. PR #380 already moved in this
  direction — it states in `skills/algorithms/skillopt/SKILL.md` that the knob is a heuristic that has
  not been isolated, that `linear` and `cosine` differ at only 2 of 12 steps over a three-value range,
  and that tuning budget is better spent on `--n-trials` and the gate.

My read: advisory is the defensible position, and the fix is mostly honesty plus deleting the schedule
options that cannot be distinguished. But that is a maintainer's call about what SkillOpt is *for*, which
is why this is filed rather than fixed.

Either way, **the telemetry should be used**: if `applied_changes > edit_budget` is worth logging, it is
worth surfacing — at minimum a warning in the run, so a reader can see the budget was ignored rather than
inferring it from two adjacent numbers.

## Acceptance criteria

- [ ] A recorded decision on whether `L` binds or is advisory.
- [ ] If binding: enforcement plus a test that a candidate exceeding `L` does not silently proceed.
- [ ] If advisory: the skill and any docs stop implying a constraint, and the indistinguishable schedule
      options are removed or justified with an ablation.
- [ ] `applied_changes > edit_budget` surfaces somewhere a human will see it, rather than only sitting in
      the event stream.

## Related

- **#371** — the mini-batch focus block this same loop renders is always empty (train ids filtered
  against val rows), so the content being annealed is also absent. The two together mean SkillOpt's
  distinctive machinery is currently neither constraining nor informed.
- **#365** — the umbrella pattern: guards that do not guard, and prompt-input defects invisible to a test
  suite that uses a prompt-ignoring optimizer.

## Provenance

Surfaced by the SkillOpt skill review (PR #380), which observed `applied_changes: 6` under `L=4` in a live
run and documented it as a known gap. Re-verified here directly against main: `skillopt.py:115` (prompt),
`:337-338` and `:344-345` (telemetry), and the absence of any comparison anywhere in the module.
